lib/rules/padding-line-between-statements.js

```diff
@@ -0,0 +1,585 @@
+    if (node.type === "DoWhileStatement" && node.body.type === "BlockStatement") {
+        return true;
+    }
+
+    // IIFE is a block-like statement specially from
+    // JSCS#disallowPaddingNewLinesAfterBlocks.
+    if (isIIFEStatement(node)) {
+        return true;
+    }
+
+    // Checks the last token is a closing brace of blocks.
+    const lastToken = sourceCode.getLastToken(node, astUtils.isNotSemicolonToken);
+    const belongingNode = astUtils.isClosingBraceToken(lastToken)
+        ? sourceCode.getNodeByRangeIndex(lastToken.range[0])
+        : null;
+
+    return Boolean(belongingNode) && (
+        belongingNode.type === "BlockStatement" ||
+        belongingNode.type === "SwitchStatement"
+    );
+}
+
+/**
+ * Check whether the given node is a directive or not.
+ * @param {ASTNode} node The node to check.
+ * @param {SourceCode} sourceCode The source code object to get tokens.
+ * @returns {boolean} `true` if the node is a directive.
+ */
+function isDirective(node, sourceCode) {
+    return (
+        node.type === "ExpressionStatement" &&
+        (
+            node.parent.type === "Program" ||
+            (
+                node.parent.type === "BlockStatement" &&
+                astUtils.isFunction(node.parent.parent)
+            )
+        ) &&
+        node.expression.type === "Literal" &&
+        typeof node.expression.value === "string" &&
+        !astUtils.isParenthesised(sourceCode, node.expression)
+    );
+}
+
+/**
+ * Check whether the given node is a part of directive prologue or not.
+ * @param {ASTNode} node The node to check.
+ * @param {SourceCode} sourceCode The source code object to get tokens.
+ * @returns {boolean} `true` if the node is a part of directive prologue.
+ */
+function isDirectivePrologue(node, sourceCode) {
+    if (isDirective(node, sourceCode)) {
+        for (const sibling of node.parent.body) {
+            if (sibling === node) {
+                break;
+            }
+            if (!isDirective(sibling, sourceCode)) {
+                return false;
+            }
+        }
+        return true;
+    }
+    return false;
+}
+
+/**
+ * Gets the actual last token.
+ *
+ * If a semicolon is semicolon-less style's semicolon, this ignores it.
+ * For example:
+ *
+ *     foo()
+ *     ;[1, 2, 3].forEach(bar)
+ *
+ * @param {SourceCode} sourceCode The source code to get tokens.
+ * @param {ASTNode} node The node to get.
+ * @returns {Token} The actual last token.
+ * @private
+ */
+function getActualLastToken(sourceCode, node) {
+    const semiToken = sourceCode.getLastToken(node);
+    const prevToken = sourceCode.getTokenBefore(semiToken);
+    const nextToken = sourceCode.getTokenAfter(semiToken);
+    const isSemicolonLessStyle = Boolean(
+        prevToken &&
+        nextToken &&
+        astUtils.isSemicolonToken(semiToken) &&
+        semiToken.loc.start.line !== prevToken.loc.end.line &&
+        semiToken.loc.end.line === nextToken.loc.start.line
```

Comment:
Will this work correctly for `EmptyStatement` nodes?

```js
if (foo) {
  bar()
}

;[1, 2, 3].map(foo)
```

It seems like the `}` will be considered the last token of the `;` node, which doesn't seem expected.